File: reactor-java/src/main/java/com/github/mostroverkhov/r2/java/R2Client.java
package com.github.mostroverkhov.r2.java;

import com.github.mostroverkhov.r2.core.*;
import com.github.mostroverkhov.r2.core.internal.acceptor.ClientAcceptor;
import com.github.mostroverkhov.r2.core.internal.requester.ClientSetup;
import com.github.mostroverkhov.r2.core.internal.requester.SetupData;
import io.rsocket.RSocket;
import io.rsocket.transport.ClientTransport;
import io.rsocket.util.PayloadImpl;
import kotlin.jvm.functions.Function1;
import org.jetbrains.annotations.NotNull;
import reactor.core.publisher.Mono;

import static io.rsocket.RSocketFactory.ClientRSocketFactory;

public class R2Client extends R2ClientFluentBuilder<
    ClientRSocketFactory,
    ClientAcceptorBuilder,
    ClientTransport,
    Mono<RequesterFactory>> {

  private Function1<
      ? super ClientAcceptorBuilder,
      ? extends ClientAcceptorBuilder> configurer;
  private Metadata metadata;
  private ClientTransport clientTransport;

  @NotNull
  @Override
  public R2ClientFluentBuilder<
      ClientRSocketFactory,
      ClientAcceptorBuilder,
      ClientTransport,
      Mono<RequesterFactory>> metadata(@NotNull Metadata metadata) {
    this.metadata = metadata;
    return this;
  }

  @NotNull
  @Override
  public R2ClientFluentBuilder<
      ClientRSocketFactory,
      ClientAcceptorBuilder,
      ClientTransport,
      Mono<RequesterFactory>> transport(ClientTransport clientTransport) {
    this.clientTransport = clientTransport;
    return this;
  }

  @NotNull
  @Override
  public R2ClientFluentBuilder<
      ClientRSocketFactory,
      ClientAcceptorBuilder,
      ClientTransport,
      Mono<RequesterFactory>> configureAcceptor(
      @NotNull Function1<
          ? super ClientAcceptorBuilder,
          ? extends ClientAcceptorBuilder> f) {
    configurer = f;
    return this;
  }

  @Override
  public Mono<RequesterFactory> start() {
    assertState();

    ClientAcceptorBuilder acceptorBuilder =
        new ClientAcceptorBuilder();

    ClientAcceptorBuilder configuredBuilder =
        configurer
            .invoke(acceptorBuilder);
    ClientAcceptor<RSocket, RSocket> acceptor =
        configuredBuilder
            .build();

    DataCodec requesterCodec = configuredBuilder
        .codecs()
        .primary();

    @SuppressWarnings("ConstantConditions")
    ClientRSocketFactory rSocketFactory = withSetup(getRSocketFactory());

    Mono<RSocket> rSocket =
        rSocketFactory
            .acceptor(acceptor::accept)
            .transport(clientTransport)
            .start();

    Mono<RequesterFactory> requesterFactory =
        rSocket
            .map(RequesterBuilder::new)
            .map(requesterBuilder -> requesterBuilder.codec(requesterCodec))
            .map(CoreRequesterBuilder::build);

    return requesterFactory;
  }

  private ClientRSocketFactory withSetup(ClientRSocketFactory factory) {
    SetupData setup = ClientSetup.metaData(metadata);
    return factory
        .dataMimeType(setup.getDataType())
        .metadataMimeType(setup.getMetadataType())
        .setupPayload(
            new PayloadImpl(
                setup.getData(),
                setup.getMetadata())
        );
  }

  private void assertState() {
    assertArg(getRSocketFactory(), "ClientRSocketFactory");
    assertArg(clientTransport, "ClientTransport");
    assertArg(configurer, "RequesterConfigurer");
  }

  private static void assertArg(Object arg, String name) {
    if (arg == null) {
      throw new IllegalArgumentException(name + " was not set");
    }
  }
}

File: rxjava-kotlin/src/main/java/com/github/mostroverkhov/r2/rxjava/R2Client.kt
package com.github.mostroverkhov.r2.rxjava

import com.github.mostroverkhov.r2.core.CoreRequesterBuilder
import com.github.mostroverkhov.r2.core.Metadata
import com.github.mostroverkhov.r2.core.R2ClientFluentBuilder
import com.github.mostroverkhov.r2.core.RequesterFactory
import com.github.mostroverkhov.r2.core.internal.requester.metaData
import io.reactivex.Single
import io.rsocket.kotlin.DefaultPayload
import io.rsocket.kotlin.RSocketFactory.ClientRSocketFactory
import io.rsocket.kotlin.transport.ClientTransport

typealias FluentBuilder = R2ClientFluentBuilder<
        ClientRSocketFactory,
        ClientAcceptorBuilder,
        ClientTransport,
        Single<RequesterFactory>>

typealias AcceptorConfigurer = (ClientAcceptorBuilder) -> ClientAcceptorBuilder

class R2Client : FluentBuilder() {
    private var configurer: AcceptorConfigurer? = null
    private var metadata: Metadata? = null
    private var clientTransport: ClientTransport? = null

    override fun metadata(metadata: Metadata): FluentBuilder {
        this.metadata = metadata
        return this
    }

    override fun transport(transport: ClientTransport): FluentBuilder {
        clientTransport = transport
        return this
    }

    override fun configureAcceptor(f: AcceptorConfigurer): FluentBuilder {
        configurer = f
        return this
    }

    override fun start(): Single<RequesterFactory> {
        assertState()
        val acceptorBuilder = ClientAcceptorBuilder()
        val transport = clientTransport!!
        val configure = configurer!!

        val configuredAcceptorBuilder = configure(acceptorBuilder)
        val clientAcceptor = configuredAcceptorBuilder.build()
        val requesterCodec = configuredAcceptorBuilder.codecs().primary()

        val rSocket = withSetup(rSocketFactory)
                .acceptor { clientAcceptor::accept }
                .transport(transport)
                .start()

        return rSocket.map(::RequesterBuilder)
                .map { it.codec(requesterCodec) }
                .map(CoreRequesterBuilder::build)
    }

    private fun assertState() {
        assertArg(clientTransport, "ClientRSocketFactory")
        assertArg(clientTransport, "ClientTransport")
        assertArg(configurer, "RequesterConfigurer")
    }

    private fun assertArg(arg: Any?, name: String) {
        if (arg == null) {
            throw IllegalArgumentException("$name was not set")
        }
    }

    private fun withSetup(factory: ClientRSocketFactory?): ClientRSocketFactory {
        val setupData = metaData(metadata)
        return factory
                ?.mimeType {
                    it.dataMimeType(setupData.dataType)
                            .metadataMimeType(setupData.metadataType)
                }
                ?.setupPayload(
                        DefaultPayload(
                                setupData.data,
                                setupData.metadata)
                )
                ?: throw IllegalArgumentException("ClientRSocketFactory not set")
    }
}